File: apps/studio/backend/app/models/world.py
from __future__ import annotations

from pydantic import BaseModel, Field, field_validator


class RegionDraft(BaseModel):
    name: str = Field(min_length=2, max_length=120)
    world_id: str = Field(default="NUM-WLD-000001")
    domain: str = Field(min_length=2, max_length=120)
    summary: str = Field(min_length=10, max_length=3000)
    educational_mission: str = Field(min_length=10, max_length=3000)
    themes: list[str] = Field(default_factory=list)
    atmosphere: list[str] = Field(default_factory=list)
    landmark_ideas: list[str] = Field(default_factory=list)

    @field_validator(
        "name",
        "world_id",
        "domain",
        "summary",
        "educational_mission",
    )
    @classmethod
    def normalize_required_text(cls, value: str) -> str:
        normalized = value.strip()

        if not normalized:
            raise ValueError("Value cannot be blank.")

        return normalized

    @field_validator(
        "themes",
        "atmosphere",
        "landmark_ideas",
    )
    @classmethod
    def normalize_lists(cls, values: list[str]) -> list[str]:
        return [
            value.strip()
            for value in values
            if value.strip()
        ]


class RegionValidationResult(BaseModel):
    valid: bool
    errors: list[str] = Field(default_factory=list)
    warnings: list[str] = Field(default_factory=list)
    proposed_id: str | None = None
    proposed_path: str | None = None


class PublishedRegion(BaseModel):
    id: str
    name: str
    world_id: str
    path: str
    message: str

File: apps/studio/frontend/src/api/world.ts
export interface RegionDraft {
  name: string;
  world_id: string;
  domain: string;
  summary: string;
  educational_mission: string;
  themes: string[];
  atmosphere: string[];
  landmark_ideas: string[];
}

export interface RegionValidationResult {
  valid: boolean;
  errors: string[];
  warnings: string[];
  proposed_id: string | null;
  proposed_path: string | null;
}

export interface PublishedRegion {
  id: string;
  name: string;
  world_id: string;
  path: string;
  message: string;
}

const API_BASE_URL = "http://127.0.0.1:8001";

async function postJson<TResponse extends object>(
  path: string,
  body: unknown,
): Promise<TResponse> {
  const response = await fetch(`${API_BASE_URL}${path}`, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify(body),
  });

  const payload = (await response.json()) as
    | TResponse
    | { detail?: string };

  if (!response.ok) {
    throw new Error(
      "detail" in payload && payload.detail
        ? payload.detail
        : `Request failed: ${response.status} ${response.statusText}`,
    );
  }

  return payload as TResponse;
}

export function validateRegion(
  draft: RegionDraft,
): Promise<RegionValidationResult> {
  return postJson<RegionValidationResult>(
    "/api/creator/world/regions/validate",
    draft,
  );
}

export function publishRegion(
  draft: RegionDraft,
): Promise<PublishedRegion> {
  return postJson<PublishedRegion>(
    "/api/creator/world/regions/publish",
    draft,
  );
}
